### PR TITLE
[JENKINS-29894] Allow to specify includesPatternFile

### PR DESCRIPTION
**Pull request by Vlatombe**
[JENKINS-29894](https://issues.jenkins-ci.org/browse/JENKINS-29894)

If provided, only the first bucket will use the excludesPatternFile, the other buckets will use includesPatternFile

@reviewbybees, @kohsuke especially.

- [x] Fix reviews
- [x] Address incompatibility ([comment](https://github.com/jenkinsci/parallel-test-executor-plugin/pull/14#discussion_r36798939))
- [x] Add regression test

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/parallel_test_executor/InclusionExclusionPattern.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package org.jenkinsci.plugins.parallel_test_executor;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * @author Vincent Latombe <vincent@latombe.net>
+ */
+public class InclusionExclusionPattern {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/parallel_test_executor/InclusionExclusionPattern.java]: Nice name :)
Some documentation would be useful

File: src/main/java/org/jenkinsci/plugins/parallel_test_executor/MultipleBinaryFileParameterFactory.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+import java.util.logging.Logger;
+
+/**
+ * Essentially a copy-paste of {@link hudson.plugins.parameterizedtrigger.BinaryFileParameterFactory} that takes a
+ * list of mappings "name -> filePattern" to generate parameters.
+ *
+ * @author Vincent Latombe <vincent@latombe.net>
+ */
+public class MultipleBinaryFileParameterFactory extends AbstractBuildParameterFactory {
+    public static class Tuple {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/parallel_test_executor/MultipleBinaryFileParameterFactory.java]: Makes sense to set a more descriptive class name. Tuple may become a problem if you decide to add another field

File: src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java [hunk 3]
```diff
@@ -36,18 +36,20 @@
...
     private String testJob;
     private String patternFile;
+    private String includesPatternFile;
     private String testReportFiles;
     private boolean doNotArchiveTestResults = false;
     private List<AbstractBuildParameters> parameters;
 
     @DataBoundConstructor
-    public ParallelTestExecutor(Parallelism parallelism, String testJob, String patternFile, String testReportFiles, boolean archiveTestResults, List<AbstractBuildParameters> parameters) {
+    public ParallelTestExecutor(Parallelism parallelism, String testJob, String patternFile, String testReportFiles, boolean archiveTestResults, List<AbstractBuildParameters> parameters, String includesPatternFile) {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java]: Please keep the original constructor to maintain the backward compatibility

File: src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java [hunk 4]
```diff
@@ -62,6 +64,10 @@ public String getPatternFile() {
         return patternFile;
     }
 
+    public String getIncludesPatternFile() {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java]: <code>@CheckForNull</code>?

**Comment by oleg-nenashev** [comment]
It would be better to submit 2 PRs next time. I'm slightly concerned regarding the coding style, because the oOracle one presumes another order of declarations. Probably it's the plugin specifics

File: src/main/java/org/jenkinsci/plugins/parallel_test_executor/InclusionExclusionPattern.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package org.jenkinsci.plugins.parallel_test_executor;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * @author Vincent Latombe <vincent@latombe.net>
```
Comment (wolfs) [inline src/main/java/org/jenkinsci/plugins/parallel_test_executor/InclusionExclusionPattern.java]: Do we really need an author tag if we use a VCS? I would favor removing all author tags in the Jenkins project.

Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/parallel_test_executor/InclusionExclusionPattern.java]: My preference as well, but we do not enforce one policy or another.

File: src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java (hunk 3 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java]: Switch to a `@DataBoundConstructor` with only the minimal required parameters and use `@DataBoundSetter` for anything with a sensible default.

File: src/main/java/org/jenkinsci/plugins/parallel_test_executor/SplitStep.java [hunk 6]
```diff
@@ -42,14 +42,14 @@ public DescriptorImpl() {
...
-    public static final class Execution extends AbstractSynchronousStepExecution<List<List<String>>> {
+    public static final class Execution extends AbstractSynchronousStepExecution<List<InclusionExclusionPattern>> {
 
         @Inject private SplitStep step;
         @StepContextParameter private Run<?,?> build;
         @StepContextParameter private TaskListener listener;
 
-        @Override protected List<List<String>> run() throws Exception {
-            return ParallelTestExecutor.findTestSplits(step.parallelism, build, listener);
+        @Override protected List<InclusionExclusionPattern> run() throws Exception {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/parallel_test_executor/SplitStep.java]: This is an incompatible change I think.

[example](https://github.com/jenkinsci/parallel-test-executor-plugin/blob/7bdb270ffe02838a36f56d4de16ebe61cd3a9556/demo/JENKINS_HOME/jobs/flow/config.xml#L12-L15)

File: src/main/java/org/jenkinsci/plugins/parallel_test_executor/MultipleBinaryFileParameterFactory.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        @Override
+        public String getDisplayName() {
+            return "Multiple Binary Files (not meant to be used)";
+        }
+    }
+
+    private static Method $setLocation;
+
+    static {
+        // work around NPE fixed in the core at 4a95cc6f9269108e607077dc9fd57f06e4c9af26
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/parallel_test_executor/MultipleBinaryFileParameterFactory.java]: That is in 1.518, so add a `TODO` comment explaining what to delete when our baseline goes past that.

**Comment by jglick** [comment]
:bug: for apparent incompatibility. Should cover it with a regression test too.

**Comment by jglick** [comment]
Same :bug: as before. I would also recommend updating the Docker demo to use the new option.

**Comment by jglick** [comment]
Specifically I expect [this line](https://github.com/jenkinsci/acceptance-test-harness/blob/c375f31e67766e0f2867d269aedc94a0379652de/src/test/java/plugins/WorkflowPluginTest.java#L126) would fail if you tried

```bash
mvn -DskipTests clean install
JENKINS_WAR=$HOME/.m2/repository/org/jenkins-ci/main/jenkins-war/1.580.1/jenkins-war-1.580.1.war LOCAL_SNAPSHOTS=true mvn -f ../acceptance-test-harness -Dtest=WorkflowPluginTest\#parallelTests test
```

Suggestion: add a new `boolean` parameter to `SplitStep` for supporting includes, defaulting to false, in which case return `List<List<String>>` as now; if set to true, return `List<InclusionExclusionPattern>`. The `help-*.html` for the new parameter can describe the format of the return value, namely that it offers `includes` and `list` properties. (You will need to mark those `InclusionExclusionPattern` getters as `@Whitelisted` for use from the Groovy sandbox, meaning a dependency on `script-security`.)

**Comment by jglick** [comment]
:bee:

File: src/main/java/org/jenkinsci/plugins/parallel_test_executor/MultipleBinaryFileParameterFactory.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+        return result;
+    }
+
+
+    @Extension
+    public static class DescriptorImpl extends AbstractBuildParameterFactoryDescriptor {
+        @Override
+        public String getDisplayName() {
+            return "Multiple Binary Files (not meant to be used)";
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/parallel_test_executor/MultipleBinaryFileParameterFactory.java]: So can we just delete the descriptor? Or hide it via `DescriptorVisibilityFilter` if not?

**Pull request merged**